lib/actions/deploy.coffee

```diff
@@ -1,229 +1,194 @@
-			catch e
-				logger.logError('Error parsing reply from remote side')
-				reject(e)
-				return
-
-			if obj.type?
-				switch obj.type
-					when 'error' then reject(new Error("Remote error: #{obj.error}"))
-					when 'success' then resolve(obj)
-					when 'status' then logger.logInfo("Remote: #{obj.message}")
-					else reject(new Error("Received unexpected reply from remote: #{data}"))
-			else
-				reject(new Error("Received unexpected reply from remote: #{data}"))
-
-		uploadRequest
-		.on('error', reject)
-		.on('data', handleMessage)
+compose = require('../utils/compose')
+
+###
+Opts must be an object with the following keys:
+
+	app: the application instance to deploy to
+	image: the image to deploy; optional
+	shouldPerformBuild
+	shouldUploadLogs
+	buildEmulated
+	buildOpts: arguments to forward to docker build command
+###
+deployProject = (docker, logger, composeOpts, opts) ->
+	_ = require('lodash')
+	doodles = require('resin-doodles')
+	sdk = require('resin-sdk').fromSharedOptions()
+
+	compose.loadProject(
+		logger
+		composeOpts.projectPath
+		composeOpts.projectName
+		opts.image
+	)
+	.then (project) ->
+		# find which services use images that already exist locally
+		Promise.map project.descriptors, (d) ->
+			# unconditionally build (or pull) if explicitly requested
+			return d if opts.shouldPerformBuild
+			docker.getImage(d.image.tag ? d.image).inspect()
+			.return(d.serviceName)
+			.catchReturn()
+		.filter (d) -> !!d
+		.then (servicesToSkip) ->
+			# multibuild takes in a composition and always attempts to
+			# build or pull all services. we workaround that here by
+			# passing a modified composition.
+			compositionToBuild = _.cloneDeep(project.composition)
+			compositionToBuild.services = _.omit(compositionToBuild.services, servicesToSkip)
+			if _.size(compositionToBuild.services) is 0
+				logger.logInfo('Everything is up to date (use --build to force a rebuild)')
+				return {}
+			compose.buildProject(
+				docker
+				logger
+				project.path
+				project.name
+				compositionToBuild
+				opts.app.arch
+				opts.app.device_type
+				opts.buildEmulated
+				opts.buildOpts
+				composeOpts.inlineLogs
+			)
+			.then (builtImages) ->
+				_.keyBy(builtImages, 'serviceName')
+		.then (builtImages) ->
+			project.descriptors.map (d) ->
+				builtImages[d.serviceName] ? {
+					serviceName: d.serviceName,
+					name: d.image.tag ? d.image
+					logs: 'Build skipped; image for service already exists.'
+					props: {}
+				}
+		.tap (images) ->
+			Promise.join(
+				sdk.auth.getUserId()
+				sdk.auth.getToken()
+				sdk.settings.get('apiUrl')
+				(userId, auth, apiEndpoint) ->
+					compose.deployProject(
+						docker
+						logger
+						project.composition
+						images
+						opts.app.id
+						userId
+						"Bearer #{auth}"
+						apiEndpoint
+						!opts.shouldUploadLogs
+					)
+			)
+	.then (release) ->
+		logger.logSuccess('Deploy succeeded!')
+		logger.logSuccess("Release ID: #{release.commit}")
```

Comment:
We shouldn't call this the release id, because that exists, and it's a different thing (which we'd like to avoid users using, but they'll end up seeing in places anyway, especially if they're using the API).

At the same time, we obviously don't want 'commit' in here since that'd make no sense. `Release hash:`?

Reply:
Yeah, that's a tricky one. I think "hash" suggests you'll get repeatedly the same value for the same input and I'd rather avoid this. I'm inclined towards "release identifier" because that's it is really, a way to refer to the release. What do you think?

Reply:
>  I think "hash" suggests you'll get repeatedly the same value for the same input 

True, that is awkward. I still think it's going to be confusing that release identifier != `release.id` though. `Release key`? `Release name`? Naming is hard.

Could just copy the UI, which puts this value in a column called 'Release', and print `Release: ` all by itself.

Reply:
Neither "key" and "name" feel ideal to me. Let's follow the UI then :)